editor/plugins/spatial_editor_plugin.cpp

```diff
@@ -1623,53 +1625,61 @@ void SpatialEditorViewport::_update_freelook(real_t delta) {
 	int key_down = Object::cast_to<InputEventKey>(ED_GET_SHORTCUT("spatial_editor/freelook_down")->get_shortcut().ptr())->get_scancode();
 	int key_speed_modifier = Object::cast_to<InputEventKey>(ED_GET_SHORTCUT("spatial_editor/freelook_speed_modifier")->get_shortcut().ptr())->get_scancode();
 
-	Vector3 velocity;
+	Vector3 direction;
 	bool pressed = false;
 	bool speed_modifier = false;
 
 	const Input &input = *Input::get_singleton();
 
 	if (input.is_key_pressed(key_left)) {
-		velocity -= right;
+		direction -= right;
 		pressed = true;
 	}
 	if (input.is_key_pressed(key_right)) {
-		velocity += right;
+		direction += right;
 		pressed = true;
 	}
 	if (input.is_key_pressed(key_forward)) {
-		velocity += forward;
+		direction += forward;
 		pressed = true;
 	}
 	if (input.is_key_pressed(key_backwards)) {
-		velocity -= forward;
+		direction -= forward;
 		pressed = true;
 	}
 	if (input.is_key_pressed(key_up)) {
-		velocity += up;
+		direction += up;
 		pressed = true;
 	}
 	if (input.is_key_pressed(key_down)) {
-		velocity -= up;
+		direction -= up;
 		pressed = true;
 	}
 	if (input.is_key_pressed(key_speed_modifier)) {
 		speed_modifier = true;
 	}
 
-	if (pressed) {
-		const EditorSettings &s = *EditorSettings::get_singleton();
-		const real_t base_speed = s.get("editors/3d/freelook_base_speed");
-		const real_t modifier_speed_factor = s.get("editors/3d/freelook_modifier_speed_factor");
+	const EditorSettings &s = *EditorSettings::get_singleton();
+	real_t inertia = s.get("editors/3d/freelook_inertia");
+	if (inertia < 0)
+		inertia = 0;
 
-		real_t speed = base_speed * cursor.distance;
-		if (speed_modifier)
-			speed *= modifier_speed_factor;
+	const real_t base_speed = s.get("editors/3d/freelook_base_speed");
+	const real_t modifier_speed_factor = s.get("editors/3d/freelook_modifier_speed_factor");
 
-		velocity.normalize();
+	real_t speed = base_speed * cursor.distance;
+	if (speed_modifier)
+		speed *= modifier_speed_factor;
 
-		cursor.pos += velocity * (speed * delta);
-	}
+	Vector3 instant_velocity = direction * speed;
+
+	// Higher inertia should increase "lag" (lerp with factor between 0 and 1)
+	// Inertia of zero should produce instant movement (lerp with factor of 1)
+	// Takes reference of 60fps for units, so that inertia of 1 gives approximate lerp factor of 0.5
+	real_t factor = 1.0 / (1.0 + inertia * delta * 60.f);
+	freelook_velocity = freelook_velocity.linear_interpolate(instant_velocity, CLAMP(factor, 0, 1));
```

Comment:
A higher delta results in a smaller factor + in a less often updated linear_interpolate().
those effects combined result in pretty different behaviour on lower fps.
lower fps -> higher delta -> smaller factor -> less often called linear_interpolate() -> slower decreasing freelook_velocity.

I was wondering why the inertia felt too bad on my pc and so wonderful in your demo video.
(simplest fix would be to allow floating point numbers for inertia in the settings. than low hardware users could use something like 0.2, bad fix I guess)

other fix:
`real_t factor = 1.0 / (1.0 + inertia * 60.f) * delta;`
this would have almost the same behaviour on both (low and high fps) on 60 fps inertia 1 we get a 1/60 of the value we get on 1 fps. (which would be correct. it still would update more often but that is a little difference)
but the 60 in the equation are not comparable to a framerate anymore than. because they just represent an arbitrary inertia factor. so we could just remove it and allow values from 1-100 for inertia. sadly this math requires us to add an `if` for the case of 0 inertia since there is now always a little bit inertia since delta is rarely 1...

I honstly would go for a completely different approach. instead of just accelerating linearly and breaking linearly. I would calculate a virtual position (base on a linear movement with a fixed speed) and than tween to it. That would have the advantage that we get smooth starting end ending points, but it is much more controllable. since the same pressing time on a key always results in the same distance moved. with the current approach there is always the additional distance of slowing down added which makes the behaviour less predictable. the distance is always the same independent if you traveled 100 or 1 unit.

I can't guarantee that I'm correct with the fix I prompted out. just was wandering why it looked so different on my machine. and thats the first thing I saw.
@Zylann would be cool if you could correct if I'm wrong otherwise I could do some test with the implementation I proposed.

Reply:
Oh my bad, it looks like I entered an integer, I should have added `.0` after it, that's why you can't enter floating point xD 

Yeah I put 60 here thinking it wasnt too bad if it's not accurate on lower fps.
You can't put delta as an end multiplier because inertia of 0 is supposed to output a factor of 1, which would be equal to delta instead, which is < 1. I think the `60` should be replaced by `1 / delta` instead.

> I honstly would go for a completely different approach. instead of just accelerating linearly and breaking linearly. I would calculate a virtual position (base on a linear movement with a fixed speed) and than tween to it.

This is what `lerp` does. When you `lerp(a, b, f)`, `b` is the target virtual position, which happens to be the expected one when inertia is disabled, hence why a factor of 1 works. You can't use a tween because start and end positions are constantly moving.

Reply:
@zylann

You are totally right about the current approach with linear_interpolate not being linear. that was dump...
I meant it is not resulting in same start and end position instead it is just creating an acceleration and a deaccelaration at the end and beginning of the button being pressed (not liner but in exponential...). So the distance traveled is not consistent. really high values for inertia result in a long path for breaking and a lot of additional distance moved.

 I'm pretty sure that using tween and updating it every frame has pretty decent results. I think I tried that once.
basically it would just use the state of the current frame and based on that it would calculate the next position.

It is also not necessary to use tween.
It also could use it's own math. I just think it would be great if pressing `D` for 1 second always results in the same new_pos independent of the inertia setting and linear to the speed settings.


I'm imagining this:

```cpp
Vector3 target_pos;
each_frame():
   if_(freelook_started()){
      target_pos = get_position();
   }
   if (w_pressed()){
      dir += forward;
   }
   //for all directions the same thing
   ...


   target_pos += dir.normalized() * speed;
   
   Vector3 offset_to_taget = (target_pos - get_position())
   Vector3 movement_this_frame = offset_to_taget.normalized() * (offset_to_taget.get_length() * (1/1+inertia);
   set_position(get_position + movement_this_frame);
```

I know this would result in infinite iterations. (without unlimited precision of floats...) but since the user releases freeload after some time anyways it will stop than again.
with this logic We would always get the same distance. smooth acceleration and smooth breaking. higher inertia would decrease the acellaration -> would make it more delayed and more smooth and lower inertia the opposite. inertia 0 would result in immediate movement. 
 - `offset_to_taget.normalized() * (offset_to_taget.get_length() * (1/1+inertia)` =
 - `offset_to_taget.normalized() * (offset_to_taget.get_length() * 1` = 
 - `offset_to_taget`
and offset_to_target is the difference between the current position and the target position. so adding offset_to_target results in current position being the target.

Reply:
I tried first with an acceleration logic, then reverted it because it was too complicated. Problem was deceleration. How do you decelerate?
I think you should create a 2D prototype to illustrate this.

Reply:
@zylann 2d prototype might be a good idea.
Ill also try to explain it from the code snippet:

 - the user stops pressing a key (`D`)
 - target_pos stops changing. (`target_pos += dir.normalized() * speed;` this is the only line where it changes and it is dependent on dir which only gets set by keys)
 - current position and the target pos are still different -> over a couple of frames the object will still get moved although there is no input anymore. (endlesslly long because it gets closer and closer to the target pos. but never reaches it. (in theroy)) until user stops forelock or again changes the target pos by pressing a button.

Reply:
I didn't create you a prototype in form of a project.
But it is easier like this it think.
A SPRITE WITH THIS SCRIPT (should be easy to set up ;) )
I made it 1d so it is as easy as it can get.
```
extends Sprite
var inertia = 0.3
var speed = 300
var target_pos = Vector2()
func _ready():
	target_pos = position

func _process(delta):
	if Input.is_key_pressed(KEY_A):
		target_pos.x -= speed * delta
	if Input.is_key_pressed(KEY_D):
		target_pos.x += speed * delta

	position = position.linear_interpolate(target_pos, clamp((1/inertia) * delta, 0, 1))
```
I kept it as easy as possible. values form 0 - 1 for inertia are good... higher works but is not really usable.

Reply:
High inertia isnt supposed to be usable anyways, we could eventually clamp it.
Will test when I get home

Reply:
cool!

Reply:
Inertia of zero yields +inf though

Reply:
I tested it, but besides the different inertia formula (which I tweaked a little), I don't see a difference with my code:
```gdscript
extends Sprite

var inertia = 0.3
var speed = 300
var target_pos = Vector2()
var velocity = Vector2()

func _ready():
	target_pos = position

func _process(delta):
	#process_by_velocity(delta)
	process_by_position(delta)

func get_dir():
	var dir = Vector2()
	if Input.is_key_pressed(KEY_LEFT):
		dir.x -= 1
	if Input.is_key_pressed(KEY_RIGHT):
		dir.x += 1
	if Input.is_key_pressed(KEY_UP):
		dir.y -= 1
	if Input.is_key_pressed(KEY_DOWN):
		dir.y += 1
	return dir

func process_by_position(delta):
	target_pos += get_dir() * speed * delta
	var f = (1.0 / (inertia + 0.001)) * delta
	position = position.linear_interpolate(target_pos, clamp(f, 0, 1))

func process_by_velocity(delta):
	var target_velocity = get_dir() * speed
	var f = (1.0 / (inertia + 0.001)) * delta
	velocity = velocity.linear_interpolate(target_velocity, clamp(f, 0, 1))
	position += velocity * delta
```

Reply:
there is not really a difference in terms of how it looks and feels.
But with process_by_position it is guaranteed that pressing the button for 1 second results in: moved by distance (speed). this has the advantage, that it is more controllable.
But you that is not that important. the more important thing is that delta is a multiplier for that factor.